Add max coinbase out sigops to CoinbaseOutputDataSize in TP protocol

Fi3 opened the pull request:
The template provide need a way to know how many sigops the pool want to include in the coinbase outupts, otherwise is not able to create a template that if mined will result in a valid block.

Sjors commented:
I can implement this on the template provider side.

Let's recommend 400 as the default, as that's what Bitcoin Core has been using for ages (`nBlockSigOpsCost = 400;` in `src/node/miner.cpp`). It can be 0 only if all coinbase payout addresses are taproot.

This will need some coordination in deployment, especially if we don't change the protocol version number (see #82).

Sjors commented on 07-Template-Distribution-Protocol.md:
```diff
@@ -12,17 +12,18 @@ Template Providers MUST attempt to broadcast blocks which are mined using work t
 | Field Name                          | Data Type | Description                                                                                     |
 | ----------------------------------- | --------- | ----------------------------------------------------------------------------------------------- |
 | coinbase_output_max_additional_size | U32       | The maximum additional serialized bytes which the pool will add in coinbase transaction outputs |
+| coinbase_output_max_sigops          | U16       | The maximum additional sigops which the pool will add in coinbase transaction outputs           |
```
Let's add a comment in the text that this implies a limit of `65535` - slightly less than the consensus maximum of `80000`. This seems fine since I can't think of a (sane) use case of putting that many sigops in the coinbase.

Sjors replied:
(this is done)

Sjors commented on 07-Template-Distribution-Protocol.md:
```diff
@@ -12,17 +12,18 @@ Template Providers MUST attempt to broadcast blocks which are mined using work t
 
 | Field Name                          | Data Type | Description                                                                                     |
 | ----------------------------------- | --------- | ----------------------------------------------------------------------------------------------- |
 | coinbase_output_max_additional_size | U32       | The maximum additional serialized bytes which the pool will add in coinbase transaction outputs |
```
This could be U24, since the max is 4,000,000.

Fi3 replied:
since we are going to use it as an u32 do not make to much sense have an u24. Also I would like to change less possible things

Sjors replied:
It saves one byte, but that's not important for a message that's only sent once. Agree it's better avoid having to cast it.

Fi3 commented:
@Sjors I addressed your comments. When we have consensus on this I can implement it in SRI and you can add it to TP. I think the most contentious point here is how to handle versioning

Sjors commented:
I'll let you know here once I have a branch with this change in it for the Template Provider. If you can open a pull request on the SRI repo with the same change, we can test if they're compatible.

Sjors commented on 07-Template-Distribution-Protocol.md:
```diff
@@ -5,24 +5,27 @@ It effectively replaces [BIP 22](https://github.com/bitcoin/bips/blob/master/bip
+It then MUST communicate it to the Template Provider via this message.
+The Template Provider MUST NOT provide `NewMiningJob` messages which would represent consensus-invalid blocks once this additional size and sigops — along with a maximally-sized (100 byte) coinbase script field — is added.
 Further, the Template Provider MUST consider the maximum additional bytes required in the output count variable-length integer in the coinbase transaction when complying with the size limits.
+Current sigops limit per block in bitcoin is 80_000. We don't want to support use cases where
```
More diplomatically:

> We're not aware of any use case where

Sjors commented on 07-Template-Distribution-Protocol.md:
```diff
@@ -5,24 +5,27 @@ It effectively replaces [BIP 22](https://github.com/bitcoin/bips/blob/master/bip
+The Template Provider MUST NOT provide `NewMiningJob` messages which would represent consensus-invalid blocks once this additional size and sigops — along with a maximally-sized (100 byte) coinbase script field — is added.
 Further, the Template Provider MUST consider the maximum additional bytes required in the output count variable-length integer in the coinbase transaction when complying with the size limits.
+Current sigops limit per block in bitcoin is 80_000. We don't want to support use cases where
+coinbase have more the 65_535 so coinbase_output_max_sigops is an U16.
```
Maybe add:

> Note that taproot outputs consume 0 sigops.

Sjors commented:
> Let's recommend 400 as the default

This is still missing.

Fi3 commented:
> > Let's recommend 400 as the default
> 
> This is still missing. Perhaps worded as:
> 
> > Use 400 for optimal backwards compatibility with v1. Note that taproot outputs consume 0 sigops.

not sure if we should recommend a default at all. If so I would go with 0.

Sjors commented:
0 will create an invalid block if you your coinbase output is non-taproot. And you won't find out until someone starts spamming the mempool with bare multisig: https://bitcoinexplainedpodcast.com/@nado/episodes/bitcoin-explained-76-stamps-and-the-invalid-block-caused-by-it

The reason I suggest 400 is because that's what `getblocktemplate` does, so every stratum v1 miner does it by default - unless they patched Bitcoin Core.

Sjors commented on 07-Template-Distribution-Protocol.md:
```diff
@@ -5,24 +5,27 @@ It effectively replaces [BIP 22](https://github.com/bitcoin/bips/blob/master/bip
 | Field Name                          | Data Type | Description                                                                                     |
 | ----------------------------------- | --------- | ----------------------------------------------------------------------------------------------- |
 | coinbase_output_max_additional_size | U32       | The maximum additional serialized bytes which the pool will add in coinbase transaction outputs |
+| coinbase_output_max_sigops          | U16       | The maximum additional sigops which the pool will add in coinbase transaction outputs           |
```
Should we add `_additional` for consistency?

plebhash replied:
```suggestion
| coinbase_output_max_additional_sigops          | U16       | The maximum additional sigops which the pool will add in coinbase transaction outputs           |
```

Fi3 commented:
> 0 will create an invalid block if you your coinbase output is non-taproot. And you won't find out until someone starts spamming the mempool with bare multisig: https://bitcoinexplainedpodcast.com/@nado/episodes/bitcoin-explained-76-stamps-and-the-invalid-block-caused-by-it
> 
> The reason I suggest 400 is because that's what `getblocktemplate` does, so every stratum v1 miner does it by default - unless they patched Bitcoin Core. This has never been configurable before.

what about: legacy pools should use 400, new pools should use 0 ?

Sjors commented:
> new pools should use 0

That assumes new pools want to use Taproot, that seems too DEMANDing :-)

Sjors commented:
Here you go, completely untested: https://github.com/Sjors/bitcoin/pull/45 (branch `sjors/sv2-sigops`)

GitGab19 commented:
@Fi3 how a JDC could know how many sigops has to signal to its TP? Shouldn't it receive them in the `AllocateMiningJobToken.Success`?

plebhash commented:
> > @Fi3 how a JDC could know how many sigops has to signal to its TP? Shouldn't it receive them in the `AllocateMiningJobToken.Success`?
> 
> if it have the outputs it can parse them I'm wondering how it communicate it to the TP btw

JDS informs JDC about `coinbase_output_max_additional_size` via `AllocateMiningJobToken.Success`

then, JDC informs TP about `coinbase_output_max_additional_size` via  `CoinbaseOutputDataSize`

---

this PR proposes introducing `coinbase_output_max_sigops` into `CoinbaseOutputDataSize`, while renaming the message into `CoinbaseOutputConstraints`

this means JDC is going to inform TP about `coinbase_output_max_sigops` via `CoinbaseOutputConstraints`

@GitGab19 question is: how is JDC informed about `coinbase_output_max_sigops`?

to me, the most logical answer simply follows the same pattern as `coinbase_output_max_additional_size`:

JDS informs JDC about `coinbase_output_max_sigops` via `AllocateMiningJobToken.Success`

---

therefore, IMO this PR should also propose adding a new `coinbase_output_max_sigops` field to `AllocateMiningJobToken.Success`

Sjors commented:
> this PR should also propose

That makes sense.

The Mining interface in Bitcoin Core already supports passing passing `coinbase_output_max_sigops` to it along with `coinbase_output_max_additional_size`: 

https://github.com/bitcoin/bitcoin/blob/97b790e844abd2f92c928239a7dc786d37fad18b/src/interfaces/mining.h#L88-L95

https://github.com/bitcoin/bitcoin/blob/97b790e844abd2f92c928239a7dc786d37fad18b/src/node/types.h#L30-L46

So this spec change won't require a change to the interface, which means there's no rush to get it finalised before the v29 Bitcoin Core release (which hopefully ships this interface).

GitGab19 commented:
> > Related to this, I realized a while ago that `NewTemplate` message contains `coinbase_outputs` field, but actually that is not set by TP, because it doesn't know the pool_outputs to add. Why couldn't we send the `coinbase_outputs` to TP, so that it can create a template with values already placed? Isn't this point related to this PR? What am I missing?
> 
> I'm pretty sure that is setted by TP. That field should contain the output with the segwit commitment (only the TP know it if we don't request txs data)

Actually it's not (at least from my findings) because they are never communicated to the TP. How could it know them by itself?
Here's the point in which the outputs are set in the jobs: https://github.com/stratum-mining/stratum/blob/73f51dba0d637f13a9dfb1fd7e4c3a7b58f59007/protocols/v2/roles-logic-sv2/src/job_creator.rs#L177

GitGab19 commented:
I did and I was not able to decode any tx output from those bytes. Can you tell me which message is used to communicate to TP the pool outputs?

GitGab19 commented:
How does it know by itself? I'm talking about the pool coinbase outputs which needs to be put into the coinbase tx

GitGab19 commented:
But the `coinbase_tx_outputs` field contained in the `NewTemplate` message is not used anywhere. The Pool (or JDC) use the coinbase_tx_outputs configured in the Pool (or the ones communicated by JDS). Am I missing something here? Shouldn't TP directly put the real coinbase_tx_outputs in that field (the one in `NewTemplate` message)?

GitGab19 commented:
> you don't see it explicitly in the roles code, cause this is something that who develop a role do not need to worry about, if I rember correctly is done somwhere in the roles-logic library

I just found this comment on the function which computes the coinbase tx: 
`/// It assume that NewTemplate.coinbase_tx_outputs == 0`

https://github.com/stratum-mining/stratum/blob/73f51dba0d637f13a9dfb1fd7e4c3a7b58f59007/protocols/v2/roles-logic-sv2/src/job_creator.rs#L323

GitGab19 commented:
I don't think so, because as I was saying earlier, the `coinbase_tx_outputs` field contained in the `NewTemplate` message is not used anywhere

GitGab19 commented:
Thank you for pointing it out, I was missing it! So it is definitely used 👍 
Why couldn't the TP directly send out the final `coinbase_outputs`? Instead of building it there? It would be a much cleaner approach imo

GitGab19 commented:
> The TP protocol is designed to be as unidirectional as possible - it doesn't think it just dumps work. It isn't fully unidirectional as it considers the size of coinbase the client wants, but that's the best we can do.
> 
> Passing the coinbase tx output set from the JDS to the TP just to have the TP echo then back to the client doesn't accomplish anything, IMO, and reduces the unidirectionality goal. It's trivial to just have the TP+JD client figure it out.

Makes sense, thanks for your answer!

Sjors commented:
> TP outputs must be placed at the end. Cause it cointains the output with the segwit commitment that must be the last output of the coinbase

I don't think it needs to be the last one, just needs to have the right magic bytes.
https://github.com/bitcoin/bips/blob/master/bip-0141.mediawiki#commitment-structure

plebhash commented on 06-Job-Declaration-Protocol.md:
```diff
@@ -38,7 +38,7 @@ Notably, if the pool intends to change the space it requires for coinbase transa
 | request_id                          | U32       | Unique identifier for pairing the response                                                                                                                                                                                                                                                                                                                  ...
 | mining_job_token                    | B0_255    | Token that makes the client eligible for committing a mining job for approval/transaction declaration or for identifying custom mining job on mining connection.                                                                                                                                                                                            ...
-| coinbase_output_max_additional_size | U32       | The maximum additional serialized bytes which the pool will add in coinbase transaction outputs. See discussion in the Template Distribution Protocol's CoinbaseOutputDataSize message for more details.                                                                                                                                                    ...
+| coinbase_output_max_additional_size | U32       | The maximum additional serialized bytes which the pool will add in coinbase transaction outputs. See discussion in the Template Distribution Protocol's CoinbaseOutputConstraints message for more details.                                                                                                                                                 ...
```
as pointed out here https://github.com/stratum-mining/sv2-spec/pull/86#issuecomment-2447119433 

we need to add `coinbase_output_max_sigops` to `AllocateMiningJobToken.Success`

plebhash commented on 06-Job-Declaration-Protocol.md:
```diff
@@ -38,7 +38,7 @@ Notably, if the pool intends to change the space it requires for coinbase transa
 | request_id                          | U32       | Unique identifier for pairing the response                                                                                                                                                                                                                                                                                                                  ...
 | mining_job_token                    | B0_255    | Token that makes the client eligible for committing a mining job for approval/transaction declaration or for identifying custom mining job on mining connection.                                                                                                                                                                                            ...
-| coinbase_output_max_additional_size | U32       | The maximum additional serialized bytes which the pool will add in coinbase transaction outputs. See discussion in the Template Distribution Protocol's CoinbaseOutputDataSize message for more details.                                                                                                                                                    ...
+| coinbase_output_max_additional_size | U32       | The maximum additional serialized bytes which the pool will add in coinbase transaction outputs. See discussion in the Template Distribution Protocol's CoinbaseOutputConstraints message for more details.                                                                                                                                                 ...
```
```suggestion
| coinbase_output_max_additional_size | U32       | The maximum additional serialized bytes which the pool will add in coinbase transaction outputs. See discussion in the Template Distribution Protocol's CoinbaseOutputConstraints message for more details.                                                                                                                                                                 |
| coinbase_output_max_additional_sigops | U16 | The maximum additional sigops which the pool will add in coinbase transaction outputs. See discussion in the Template Distribution Protocol's CoinbaseOutputConstraints message for more details. |
```

plebhash replied:
as pointed out here https://github.com/stratum-mining/sv2-spec/pull/86#issuecomment-2447119433 

we also need to add `coinbase_output_max_additional_sigops` to `AllocateMiningJobToken.Success`

Sjors commented:
I rebased https://github.com/Sjors/bitcoin/pull/45, so you can test it against a modified SRI.

Sjors commented:
I added a commit that makes the Template Provider backwards compatible, it will just assume max additional sigops is 400 if the field is missing.

Suggested deployment strategy:

1. Approve, but not merge, this spec change PR
2. I'll make a new Template Provider release which includes https://github.com/Sjors/bitcoin/pull/45
3. Update SRI to set the field
4. Release new SRI version (maybe have instructions specify a minimum TP release)
5. Merge this spec change

Sjors commented:
Should I go ahead and include https://github.com/Sjors/bitcoin/pull/45 in the next TP release? Probably in the next few weeks.

GitGab19 commented:
Sounds good to me 👍

Sjors commented:
Done. I'll let you know when the new release is out.

Conversely, please let me know which release of SRI will have this field, so I can add a code comment about that.

GitGab19 commented:
Will add this as a discussion point of our next dev call and get back to you.

Sjors commented:
It's fine to just ping me once it happens, no need to plan in advance.

GitGab19 commented on 08-Message-Types.md:
```diff
@@ -42,7 +42,7 @@
 | 0x60                           | 1               | SubmitSolution                     |
 | Template Distribution Protocol |                 |                                    |
-| 0x70                           | 0               | CoinbaseOutputDataSize             |
+| 0x70                           | 0               | CoinbaseOutputConstraint           |
```
```suggestion
| 0x70                           | 0               | CoinbaseOutputConstraints           |
```

GitGab19 reviewed (commented):
Just found a typo @plebhash

Sjors commented:
The new release https://github.com/Sjors/bitcoin/releases/tag/sv2-tp-0.1.14 supports the new `coinbase_output_max_additional_sigops` field. Binaries coming in the next few hours.

Sjors commented:
Nice! Just let me know when there is a new SRI release, so I can also make a note in the TP code.
